cli-plugins/manager/manager.go

```diff
@@ -0,0 +1,190 @@
+package manager
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/spf13/cobra"
+)
+
+// ErrPluginNotFound is the error returned when a plugin could not be found.
+type ErrPluginNotFound string
+
+func (e ErrPluginNotFound) Error() string {
+	return "Error: No such CLI plugin: " + string(e)
+}
+
+var (
+	pluginDirs     []string
+	pluginDirsOnce sync.Once
+)
+
+func getPluginDirs(dockerCli command.Cli) []string {
+	pluginDirsOnce.Do(func() {
+		// Mostly for test.
+		if ds := os.Getenv("DOCKER_CLI_PLUGIN_EXTRA_DIRS"); ds != "" {
+			pluginDirs = append(pluginDirs, strings.Split(ds, ":")...)
+		}
+
+		if cfg := dockerCli.ConfigFile(); cfg != nil {
+			pluginDirs = append(pluginDirs, cfg.CLIPluginExtraDirs...)
+		}
+
+		pluginDirs = append(pluginDirs, defaultPluginDirs...)
+	})
+	return pluginDirs
+}
+
+func addPluginCandidatesFromDir(res map[string][]string, d string) error {
+	dentries, err := ioutil.ReadDir(d)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		// Portable? This seems to be as good as it gets :-/
+		if serr, ok := err.(*os.SyscallError); ok && serr.Err == syscall.ENOTDIR {
```

Comment:
What about checking if the entry is a directory BEFORE the `ReadDir` ?
```go
 if info, err := os.Stat(d); err != nil || !info.IsDir() {
       return fmt.Errorf("failed to stat %q", d) 
 }
```

Reply:
Why didn't I think of that! (Perhaps trying to avoid ToCToU issues, but why!?!). I'll do as you suggest.